File: core/combat.py
from typing import Callable, Iterable

from core.attack import AttackProfile, AttackState
from core.damage import calc_damage


class TickController:
    """Accumulates time and invokes a callback on fixed ticks.

    Example:
        tc = TickController(0.2)
        tc.update(0.1, on_tick)  # no tick yet
        tc.update(0.1, on_tick)  # on_tick(0.2) called once
    """

    def __init__(self, tick_length_s: float = 0.2) -> None:
        self.tick_length_s = float(tick_length_s)
        self._accum = 0.0

    def update(self, dt: float, on_tick: Callable[[float], None]) -> None:
        self._accum += float(dt)
        while self._accum >= self.tick_length_s:
            on_tick(self.tick_length_s)
            self._accum -= self.tick_length_s

    def __str__(self) -> str:
        return (
            "TickController("
            f"dt={self.tick_length_s}s, accum={self._accum:.3f})"
        )


class CombatSystem:
    """Coordinates basic attacks each tick for a set of actors vs. one enemy.

    Expected shape for an actor and enemy:
      - .stats.atk / .stats.defense
      - .health.current with .clamp()
      - .mana.current with .clamp()
      - .attack_profile.cooldown_s / .mp_gain_on_attack
      - .attack_state.tick(dt) / .ready(cooldown_s) / .reset()

    NOTE: Fill in the TODOs to wire the behavior.
    """

    def __init__(self, actors: Iterable, enemy) -> None:
        self.actors = list(actors)
        self.enemy = enemy

    def on_tick(self, dt: float) -> None:
        """Advance attack timers and perform basic attacks when ready."""
        if self.enemy.health.is_dead():
            return  # Enemy already defeated
        for actor in self.actors:
            if actor.health.is_dead():
                continue  # Skip dead actors
            actor.attack_state.tick(dt)
            if actor.attack_state.ready(actor.attack_profile.cooldown_s):
                self.basic_attack(actor, self.enemy)


    def basic_attack(self, attacker, defender) -> int:
        """Compute damage and apply it to defender; grant attacker MP.
        Returns the damage dealt.
        """
        damage = calc_damage(attacker.stats.atk, defender.stats.defense)
        if attacker.mana.full():
            damage += attacker.magic_damage
            attacker.mana.current = 0
        defender.health.current -= damage
        defender.health.clamp()
        attacker.mana.current += attacker.attack_profile.mp_gain_on_attack
        attacker.mana.clamp()
        attacker.attack_state.reset()
        return damage

    def __str__(self) -> str:
        enemy_name = type(self.enemy).__name__
        return f"CombatSystem(actors={len(self.actors)}, enemy={enemy_name})"

File: web/src/domain/combat.ts
import { calcDamage } from "./damage";
import { Character } from "./entities";

export class TickController {
  tickLengthS: number;
  private accum: number;

  constructor(tickLengthS = 0.2) {
    this.tickLengthS = Number(tickLengthS);
    this.accum = 0.0;
  }

  update(dt: number, onTick: (dt: number) => void): void {
    this.accum += Number(dt);
    while (this.accum >= this.tickLengthS) {
      onTick(this.tickLengthS);
      this.accum -= this.tickLengthS;
    }
  }

  toString(): string {
    return `TickController(dt=${this.tickLengthS}s, accum=${this.accum.toFixed(3)})`;
  }
}

export class CombatSystem {
  actors: Character[];
  enemy: Character;
  private selectActorTarget: (actor: Character) => Character | null;
  private selectEnemyTarget: (enemy: Character) => Character | null;

  constructor(params: {
    actors: Character[];
    enemy: Character;
    selectActorTarget?: (actor: Character) => Character | null;
    selectEnemyTarget?: (enemy: Character) => Character | null;
  }) {
    this.actors = params.actors;
    this.enemy = params.enemy;
    this.selectActorTarget = params.selectActorTarget ?? ((_) => this.enemy);
    this.selectEnemyTarget =
      params.selectEnemyTarget ?? ((_) => this.actors.find((actor) => !actor.health.isDead()) ?? null);
  }

  onTick(dt: number): void {
    this.actors.forEach((actor) => {
      if (actor.health.isDead()) return;
      actor.attackState.tick(dt);
      if (!actor.attackState.ready(actor.attackProfile.cooldownS)) return;
      const target = this.selectActorTarget(actor);
      if (!target || target.health.isDead()) return;
      this.basicAttack(actor, target);
    });

    if (this.enemy.health.isDead()) return;

    this.enemy.attackState.tick(dt);
    if (!this.enemy.attackState.ready(this.enemy.attackProfile.cooldownS)) return;
    const target = this.selectEnemyTarget(this.enemy);
    if (!target || target.health.isDead()) return;
    this.basicAttack(this.enemy, target);
  }

  basicAttack(attacker: Character, defender: Character): number {
    let damage = calcDamage(attacker.stats.atk, defender.stats.defense);
    const mana = (attacker as { mana?: { full(): boolean; current: number; clamp(): void; max: number } }).mana;
    if (mana && mana.full()) {
      damage += (attacker as { magicDamage?: number }).magicDamage ?? 0;
      mana.current = 0;
    }
    defender.health.current -= damage;
    defender.health.clamp();
    if (mana) {
      mana.current += attacker.attackProfile.mpGainOnAttack;
      mana.clamp();
    }
    attacker.attackState.reset();
    return damage;
  }

  toString(): string {
    const enemyName = (this.enemy as { name?: string }).name ?? "Enemy";
    return `CombatSystem(actors=${this.actors.length}, enemy=${enemyName})`;
  }
}
